x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -0,0 +1,260 @@
+        final String relayState = parameters.get("RelayState");
+        final String sigAlg = parameters.get("SigAlg");
+        final String signature = parameters.get("Signature");
+        if (null == samlRequest) {
+            logAndThrow(() -> new ParameterizedMessage("Query string [{}] does not contain a SAMLRequest parameter", queryString));
+        }
+        AuthnRequest authnRequest = null;
+        // We consciously parse the AuthnRequest before we validate its signature as we need to get the Issuer, in order to
+        // verify if we know of this SP and get its credentials for signature verification
+        final Element root = parseSamlMessage(inflate(decodeBase64(samlRequest)));
+        if ("AuthnRequest".equals(root.getLocalName()) == false
+            || "urn:oasis:names:tc:SAML:2.0:protocol".equals(root.getNamespaceURI()) == false) {
+            logAndThrow(() -> new ParameterizedMessage("SAML message [{}] is not an AuthnRequest", SamlUtils.toString(root, true)));
+        }
+        try {
+            authnRequest = SamlUtils.buildXmlObject(root, AuthnRequest.class);
+        } catch (Exception e) {
+            logAndThrow(() -> new ParameterizedMessage("Cannot process AuthnRequest [{}]", SamlUtils.toString(root, true)), e);
+        }
+        final SamlServiceProvider sp = getSpFromIssuer(authnRequest.getIssuer());
+        if (signature != null) {
+            if (sigAlg == null) {
+                logAndThrow(() ->
+                    new ParameterizedMessage("Query string [{}] contains a Signature but SigAlg parameter is missing",
+                        queryString));
+            }
+            final X509Credential spSigningCredential = sp.getSigningCredential();
+            if (spSigningCredential == null) {
+                logAndThrow(
+                    "Unable to validate signature of authentication request, Service Provider hasn't registered signing credentials");
+            }
+            validateSignature(samlRequest, sigAlg, signature, sp.getSigningCredential(), relayState);
+        }
+        validateAuthnRequest(authnRequest, sp);
+        Map<String, Object> additionalData = buildAdditionalData(authnRequest, sp);
+        return new SamlValidateAuthnRequestResponse(sp.getEntityId(), authnRequest.isForceAuthn(), additionalData);
+    }
+
+    private Map<String, Object> buildAdditionalData(AuthnRequest request, SamlServiceProvider sp) {
+        Map<String, Object> data = new HashMap<>();
+        final NameIDPolicy nameIDPolicy = request.getNameIDPolicy();
+        if (null != nameIDPolicy) {
+            final String requestedFormat = request.getNameIDPolicy().getFormat();
+            if (Strings.hasText(requestedFormat)) {
+                data.put("nameid_format", requestedFormat);
+                // we should not throw an error. Pass this as additional data so that the /saml/init API can
+                // return a SAML response with the appropriate status (3.4.1.1 in the core spec)
```

Comment:
Are you intentionally saying that we would go through the user auth phase and _then_ return an error to the SP?
Is that necessary?

I think we should indicate to the caller that validation failed, and either:
1. return all the info they need to send back to the SP (but that seems like it duplicates the other API), _OR_
2. provide a new API that takes an error object from this response and turns it an Authn `Response` failure, _OR_
3. make it possible to call the `Response` API without user authc, if-and-only-if the state data indicates an error.

Reply:
> Are you intentionally saying that we would go through the user auth phase and then return an error to the SP?
Is that necessary?

Yes. If the user can't authenticate to the IDP we should not return `urn:oasis:names:tc:SAML:2.0:status:InvalidNameIdPolicy` but `urn:oasis:names:tc:SAML:2.0:status:AuthnFailed` . Specific statues are optional in the spec and we can select to ignore them but it's very useful for debugging and important for interoperability. 

> return all the info they need to send back to the SP (but that seems like it duplicates the other API)

That would require the caller of the API to be able to construct SAML Responses and I don't think we should expect that. 

Between 2 and 3 I was leaning towards 3 and that we would handle that in the Transport action of the other API, so that we don't need to build additional logic in the UI/caller of the API. I'm ofcourse open to discuss this further